packages/core/src/PlayerContextProvider.js

```diff
@@ -648,7 +657,13 @@ export class PlayerContextProvider extends Component {
     const { cycle, activeTrackIndex } = this.state;
     if (!cycle && activeTrackIndex + 1 >= playlist.length) {
       if (loadFirstTrackOnPlaylistComplete) {
-        this.goToTrack({ index: 0, shouldPlay: false, shouldForceLoad: true });
+        const currentTime = playlist[0].startingTime || 0;
+        this.goToTrack({
+          index: 0,
+          currentTime,
```

Comment:
This seems weird to me. It brings me back to wondering if playlist tracks are the best place to store the starting time... it seems like temporal information, not something you would want to store in a playlist. I don't think you would want to reload a track that was already played at the starting time from before.. would you?

This isn't a rhetorical question, I'm really not sure.

Reply:
Yes, you wouldn't want to reload a track at a startingTime after it ends, you could remove it from the track after it changes. However, I'm open to suggestions with regards to the API.

Reply:
I didn't consider the case where you have more than 1 source per track.

Reply:
I'm not sure what you're saying would happen with multiple sources.. I'm mostly just thinking about when you cycle back to a track that has already played. I don't want to create a bunch of extra work for the user in order to use the `startingTime` API correctly.

Reply:
Yeah, multiple sources comment isn't related to what you mentioned, it was just a thought that current API proposal doesn't support different starting tracks per source.

Reply:
Multiple sources for the same track should always be assumed you be identical content, and I'm making the assumption your first source for a given track will always be the same, so that doesn't worry me too much.

Reply:
Ohh, got it, no problem with that then

Reply:
Let's see if we can come up with something better related to the API, I'm going to share how I'm handling the issue you mentioned.

FYI: In OwlTail, the track that is being played is always the first track in the playlist, so I modify the playlist accordingly on certain actions, and use `onActiveTrackUpdate` to set the new track as the first one.

I currently don't have the issue to be fair, as I remove the track from the playlist after it ends 😅, but the same logic to remove it could be applied to set startingTime back to 0 or something like that (with current API).

When I get an `onActiveTrackUpdate`, I pass the current track to 2nd position, and the new track to the first one, however I don't know if it was because the track ended or it changed, that's why I had to use the media ref to subscribe to the ended event and set a state variable like `awaitingNextTrack`, which changes `onActiveTrackUpdate` behaviour to remove the current track instead of moving it (or setting back to startingTime: 0).

So yes, I agree that can be considered a bunch of extra work for the user and if we can change the API to make it easier, that's good. I don't think however that cassette should manage the `startingTime`, just provide good helpers to let you manage it yourself, kind of like the playlist.

Reply:
Another pain point with `onActiveTrackUpdate` and saving the startingTime in playlist only after track change actions, to prevent updating the playlist all the time, is that `onActiveTrackUpdate` it's called after `setMediaElementSources` is called, so you don't have a chance to read the currentTime of last track, as the media element is already reset. I had to make a wrapper of `onSelectTrackIndex` that calls a function to update the startingTime first and then calls `onSelectTrackIndex`.

Extra work as well.

Reply:
One thought I had to is to have something called `intialStartingTime` (instead `startingTime`) and we would keep it either in the track object or a separate prop. And it would only be respected the first time we play the track.. after that, it starts at the beginning as normal.

The problem with this is that if we save a currentTime for a skipped track then come back to it without unmounting the component, we won't have a way to tell that track to resume at the previous currentTime.

So just for a moment, considering my previous idea - where we save the currentTime for *all* tracks in the state snapshot (and assuming we implement the feature where you can swap in a new state snapshot), what about that doesn't work for you? I'm trying to understand the drawbacks of that approach.

Reply:
I think a glaring problem about that approach is inconsistencies if you have a user listening on multiple devices concurrently, but besides that.

Reply:
You should be able to switch between tracks multiple times and resume where you left (first time or not) and for that to work, I guess we would have to pass that initialStartingTime prop after every snapshot, it would be the same as updating the playlist all the time.

Reply:
I guess we could do something like that, but it wasn't my first option to keep the library simpler, since as you said, it's an advanced feature.

Reply:
I think if we were to go with the snapshot we would also override the startingTimes internally without worrying about a prop. But like I said below, I think we should go with what you've already implemented.